### ISSUE TITLE
I want to integrate cellpose as a ImagePy plugin

### ISSUE THREAD

**Issue by yxdragon** [MouseLand/cellpose#11]
fantastic work! I want to integrate it in [ImagePy](https://github.com/Image-Py/imagepy), which supports all the interaction cellpose needed. 

So can you show the core step without ui? such as read the image and label (or contours), predict and show with matplotlib?

1. load model, estimate radius?
2. give the image and the label mask (or some contours vector?)
3. predict segmentation result, a label? and use skimage's find_contours, then plot?

**Comment by carsen-stringer** [comment]
awesome! ImagePy looks great! Here's the full code to do what you're asking

~~~
from cellpose import models, utils, plot

# check if GPU working, and if so use it
use_gpu = utils.use_gpu()
if use_gpu:
    device = mx.gpu()
else:
    device = mx.cpu()

# DECLARE MODEL
# model_type='cyto' or model_type='nuclei'
model = models.Cellpose(device, model_type='cyto')

imgs = [img] # needs to be a list

# define CHANNELS to run segementation on
# grayscale=0, R=1, G=2, B=3
# channels = [cytoplasm, nucleus]
# if NUCLEUS channel does not exist, set the second channel to 0
channels = [2,3]

# RUN CELLPOSE
masks, flows, styles, diams = model.eval(imgs, rescale=None, channels=channels)

# masks are 0,1,2... representation
# PLOT outlines on image
outpix = plot.plot_outlines(masks[0]) # compute outlines
plt.imshow(imgs[0])
for k in range(len(outpix)):
    plt.plot(outpix[k][:,0], outpix[k][:,1], color='r')
plt.show()
~~~

note we have two models, 'cyto' and 'nuclei'. there's also the option of choosing the channels that is in our GUI.

please let me know if you need any more info. also here's a link to the example [notebook](https://github.com/MouseLand/cellpose/blob/master/notebooks/run_cellpose.ipynb) which has more potential plot views.

**Comment by yxdragon** [comment]
ImagePy supports multi-channels image sequence. It seems that model.eval supports sequence, So:
is "sequence eval" faster than for "loop eval one by one"? does it support a process callback function? It is important if we eval many images at once.

**Comment by yxdragon** [comment]
a tips: I saw plot.py use many numpy method to process the label. skimage has many util method, such as find_contours, make a disk, circle, label, overlay...

for example:
```python
unq = np.unique(Y)
nmask = len(unq)-1
    for j in range(nmask):
        mask = (Y==unq[j+1])
```
if the image is large and there are thousands of labels: Y==lab would melloc the image many times. that would cost much memory, and be very slow. 

so your plot.py could be more clean and utility. But it is not the core of cellpose, so I'd like to know if you need a pr for the post-processing?

**Comment by yxdragon** [comment]
Now it is OK, [here](https://github.com/Image-Py/cellpose-plgs)

quesstions:
1. I saw the qt's ui, we can set the diams manually, but in where to pass this parameter?
2. Is there a listmodel() method, return all the models, and if it is downloaded. such as [('cyto', True), ('nulei', False)]. and can I download(name) the specific model?

ImagePy has a powerful label panel (with an unsupervised AI Brush).
[here](http://skimgplgs.imagepy.org/coins_label.mp4) is a demo video, It is very useful for label.
Can you give me an interface to upload the label and data set?
I want to add "list models, install models, upload label set" in ImagePy.

**Comment by carsen-stringer** [comment]
Very cool plugin that you've made! Apologies for not responding sooner as I have been travelling.

1. Running in a loop versus running a list input should take a similar amount of time if the user is running with the standard settings, so feel free to interface with it in whatever way is easier for ImagePy.

2. I found that the built-in skimage functions were not doing well for finding borders when masks are touching. What function would you recommend? What seemed to me to be the fastest way to compute accurate boundaries was to find boundary pixels based on neighbors (see masks_to_outlines in [plot.py](cellpose/plot.py)), and this is the function that the GUI uses for display purposes.

3. I've put the model URLs below and you can call the function download_model_weights(urls) in [models.py](cellpose/models.py) for downloading. I guess I can see why you may want to separate these steps, but I don't see any point in subdividing into the two models further (the files aren't that big). If you have users of ImagePy that want to upload their images, let me know and I will expose the save_server function (right now it's a method of the main GUI).

~~~
urls = ['http://www.cellpose.org/models/cyto_0',
        'http://www.cellpose.org/models/cyto_1',
        'http://www.cellpose.org/models/cyto_2',
        'http://www.cellpose.org/models/cyto_3',
        'http://www.cellpose.org/models/size_cyto_0.npy',
        'http://www.cellpose.org/models/nuclei_0',
        'http://www.cellpose.org/models/nuclei_1',
        'http://www.cellpose.org/models/nuclei_2',
        'http://www.cellpose.org/models/nuclei_3',
        'http://www.cellpose.org/models/size_nuclei_0.npy']
~~~

**Comment by yxdragon** [comment]
1. the first  method, we can use scipy.ndimage.find_object to get every label region's slice, then just copy the slices, do as now.       just pseudo code
```python
slices = scipy.ndimage.find_object(img)
for sr, sc in slices:
      msk = img[sr, sc] == i
      pts = find_contours(msk) + [sr[0], sc[0]]   # add the offset
```
but there is a better way here:
2. we can use skimage.segmentation.find_boundaries to mark the label's edge, then invert the image, find the inner contours (that is the object's out contours). here is the code:

```python
mark[find_boundaries(mark, mode='inner')] = 0
mark[:,[0,-1]], mark[[0,-1],:] = 0, 0   # if we want leave the mark touch the border
contours, hierarchy = cv2.findContours((mark==0).astype(np.uint8), cv2.RETR_CCOMP, cv2.CHAIN_APPROX_NONE)
contours = [contours[i] for i in np.where(hierarchy[0,:,3]>0)[0]]
```
tested on the image below
![image](https://user-images.githubusercontent.com/24822467/75746459-49347780-5d55-11ea-9326-98a0aba34f54.png)

and there are many other method in plot.py could be optimized, if you need help, let me know.

**Comment by carsen-stringer** [comment]
Thanks for the help with this, `find_boundaries` works great and is super fast. I've added it to the pipeline.

Also I've changed the cellpose pip so that `pip install cellpose` does not install the graphical dependencies.
